[DNR] requestDomains/excludedRequestDomains should operate on the domain of the network request, not the domain of the loaded page

elijahsawyers opened the pull request:
#### 1915cc5dfc4b0301a2ac9da6ecb239bb9b478dd1
<pre>
[DNR] requestDomains/excludedRequestDomains should operate on the domain of the network request, not the domain of the loaded page
<a href="https://bugs.webkit.org/show_bug.cgi?id=289205">https://bugs.webkit.org/show_bug.cgi?id=289205</a>
<a href="https://rdar.apple.com/113878444">rdar://113878444</a>

Reviewed by Timothy Hatcher.

This patch makes it so that requestDomains and excludedRequestDomains operate on the
domain of the network request.

To do this, we need to include the requestDomains and excludedRequestDomains in the
url-filter of the WebKit rule that we convert the dNR rule into. However, we can&apos;t
include all domains in the url-filter because the regex doesn&apos;t allow for complex
rules (such as the | operator). To get around this, we create one WebKit rule per
request domain, and we combine each request domain with the urlFilter (if there is
one). Additionally, we create one ignore-previous-rules rule for each excluded
request domain.

This patch also uses the new top-document and child-document resource types to stop
having to create multiple rules if the dNR rule doesn&apos;t specify resource types.

It&apos;s also worth noting that if a dNR rule has requestDomains and a regexFilter, we
throw away the requestDomains.

* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
(WebKit::if):
(-[_WKWebExtensionDeclarativeNetRequestRule ruleInWebKitFormat]):
(-[_WKWebExtensionDeclarativeNetRequestRule _convertRulesWithModifiedCondition:webKitActionType:chromeActionType:]):
(-[_WKWebExtensionDeclarativeNetRequestRule _webKitRuleWithWebKitActionType:chromeActionType:condition:]):
(-[_WKWebExtensionDeclarativeNetRequestRule _chromeResourceTypeToWebKitResourceType]):
(-[_WKWebExtensionDeclarativeNetRequestRule _allChromeResourceTypesForCondition:]):
(-[_WKWebExtensionDeclarativeNetRequestRule _findLongestCommonSubstringWithString:andString:]):
(-[_WKWebExtensionDeclarativeNetRequestRule _combineRequestDomain:withURLFilter:]):
(-[_WKWebExtensionDeclarativeNetRequestRule _convertedRulesForWebKitActionType:chromeActionType:]): Deleted.
(-[_WKWebExtensionDeclarativeNetRequestRule _webKitRuleWithWebKitActionType:chromeActionType:chromeResourceTypes:]): Deleted.
(-[_WKWebExtensionDeclarativeNetRequestRule _chromeResourceTypeToWebKitLoadContext]): Deleted.
(-[_WKWebExtensionDeclarativeNetRequestRule _allChromeResourceTypes]): Deleted.
Modify the dNR to WebKit rule conversion logic to work as described in the summary.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, MainFrameResourceRuleConversion)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, SubFrameResourceRuleConversion)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RepeatedMainFrameResourceRuleConversion)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, ConvertingMultipleResourceTypes)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, ConvertingXHRWebSocketAndOtherTypes)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, UpgradeSchemeForMainFrameRuleConversion)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithDomainType)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithNoSpecifiedResourceTypes)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithUnsupportedExcludedResourceTypes)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithEmptyExcludedResourceTypes)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithURLFilterAndRequestDomains)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithRequestDomains)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithExcludedRequestDomains)):
Modify the rule conversion tests to pass, and write a new test,
RuleConversionWithURLFilterAndRequestDomains, to ensure that
rules that combine request domains and URL filters converts
as expected.

Canonical link: <a href="https://commits.webkit.org/292528@main">https://commits.webkit.org/292528@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/44ae8769399ac1c2fa68ac6a4b2adb0d2bed8581

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96227 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15841 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5809 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101293 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46746 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98272 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16137 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24274 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73357 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30588 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99230 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12110 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86932 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53696 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11861 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4693 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46072 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81983 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4790 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103320 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23294 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16964 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82397 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23545 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82951 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81773 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26395 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3835 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16668 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15503 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23257 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28412 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22916 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26396 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24657 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebCore/loader/ResourceLoadInfo.h:
```diff
@@ -55,9 +55,11 @@ enum class ResourceType : uint32_t {
-    CSPReport = 0x10000,
+    CSPReport = 0x100000,
+    TopDocument = 0x200000,
+    ChildDocument = 0x400000,
```
Since we are changing the version number, we can group these bits back together. I sent you a version of this file, since there is more to change than just these values.
```suggestion
    CSPReport = 0x1000,
    TopDocument = 0x2000,
    ChildDocument = 0x4000,
```

xeenon commented on Source/WebCore/loader/ResourceLoadInfo.h:
```diff
@@ -35,12 +35,12 @@ namespace WebCore::ContentExtensions {
     UnlessFrameURL = 0x80000
 };
-static constexpr uint32_t ActionConditionMask = 0xE0000;
+static constexpr uint32_t ActionConditionMask = 0xF0000;
```
This should stay as it was. They are mutually exclusive values, so they can overlap.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -721,61 +729,59 @@ - (void)removeInvalidResourceTypesForKey:(NSString *)ruleConditionKey
+    if (condition[ruleConditionInitiatorDomainsKey] && condition[ruleConditionExcludedInitiatorDomainsKey]) {
         // If a rule specifies both initiatorDomains and excludedInitiatorDomains, we need to turn that into two rules. The first rule will have the excludedInitiatorDomains, and be implemented
         // as an ignore-previous-rules using if-frame-url (instead of unless-frame-url).
         // To do this, make a copy of the condition dictionary, make the initiatorDomains be the excludedInitiatorDomains of the original rule, and create an ignore-previous-rules rule.
```
I think this comment is out of date, `ignore-previous-rules` comes second (I think).

elijahsawyers replied:
Yeah, you're right. The `ignore-previous-rules` rule comes second. I can update this comment and remove the second one in `_webKitRuleWithWebKitActionType:chromeActionType:condition:`. The comments are more likely to get out of date if there are two saying basically the same thing.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -893,29 +884,18 @@ - (NSDictionary *)_webKitRuleWithWebKitActionType:(NSString *)webKitActionType c
         triggerDictionary[@"unless-frame-url"] = mapObjects(excludedDomains, convertToURLRegexBlock);
 
-
     // FIXME: <rdar://72203692> Support 'allowAllRequests' when the resource type is 'sub_frame'.
```
I believe this is fixed or can be fixed now that you added `child-document`.

elijahsawyers replied:
I can take this radar and investigate it.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -986,6 +966,88 @@ - (NSArray *)_convertedResourceTypesForChromeResourceTypes:(NSArray *)chromeReso
+    // E.g.
+    // requestDomain = com, urlFilter = /foo/bar, result = com*/foo/bar
+    // requestDomain = com, urlFilter = &foo=bar|, result = com*&foo=bar|
+    return [[requestDomain stringByAppendingString:@"*"] stringByAppendingString:urlFilter];
```
```suggestion
    return [@"||" stringByAppendingString:[[requestDomain stringByAppendingString:@"*"] stringByAppendingString:urlFilter]];
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
```diff
@@ -1878,6 +1834,60 @@
     EXPECT_NS_EQUAL(convertedRule, correctRuleConversion);
 }
 
+TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithURLFilterAndRequestDomains)
```
You should add an `excludedRequestDomains` test too.

elijahsawyers replied:
We cover that with `RuleConversionWithExcludedRequestDomains`. Is there another test case that you had in mind?

xeenon replied:
Ah, just missed it.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -986,6 +966,88 @@ - (NSArray *)_convertedResourceTypesForChromeResourceTypes:(NSArray *)chromeReso
+{
+    // If we don't have a URL filter, just return the request domain; there's nothing to combine here.
+    if (!urlFilter)
+        return requestDomain;
```
```suggestion
        return [@"||" stringByAppendingString:requestDomain];
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -986,6 +966,88 @@ - (NSArray *)_convertedResourceTypesForChromeResourceTypes:(NSArray *)chromeReso
+    // E.g.
+    // requestDomain = com, urlFilter = .com/foo/bar, result = .com/foo/bar
+    if ([urlFilter containsString:requestDomain])
+        return urlFilter;
```
```suggestion
    // Remove any leading periods from the urlFilter before checking the prefix.
    NSString *trimmedFilter = urlFilter;
    while ([trimmedFilter hasPrefix:@"."])
        trimmedFilter = [trimmedFilter substringFromIndex:1];

    if ([trimmedFilter hasPrefix:requestDomain])
        return [@"||" stringByAppendingString:urlFilter];
```

elijahsawyers replied:
We should probably also do the same for the request domain. For example, if the request domain is `.com` and the url filter is `.com/foo/bar=baz`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -986,6 +966,88 @@ - (NSArray *)_convertedResourceTypesForChromeResourceTypes:(NSArray *)chromeReso
+    // requstDomain = foo.com, urlFilter = bar-*.com^, result = bar-*foo.com^
+    if (NSString *longestCommonSubstring = [self _findLongestCommonSubstringWithString:requestDomain andString:urlFilter]) {
+        if (longestCommonSubstring.length > 1)
+            return [urlFilter stringByReplacingCharactersInRange:[urlFilter rangeOfString:longestCommonSubstring] withString:requestDomain];
```
We might want to prepend `||` here too.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -721,61 +729,60 @@ - (void)removeInvalidResourceTypesForKey:(NSString *)ruleConditionKey
 
+    if (condition[ruleConditionInitiatorDomainsKey] && condition[ruleConditionExcludedInitiatorDomainsKey]) {
+        // If a rule specifies both initiatorDomains and excludedInitiatorDomains, we need to turn that into two rules.
+        // The first rule will have the initiatorDomains implemented as normal (using if-frame-url). We then create a
+        // second rule as an ignore-previous-rules rule using if-frame-url instead of unless-frame-url.
```
```suggestion
        // The first rule will have the initiatorDomains implemented as normal (using if-frame-url). We then create a
        // second rule as an ignore-previous-rules rule using unless-frame-url instead of if-frame-url.
```

elijahsawyers replied:
It's an `if-frame-url` trigger for the excluded initiator domains if we also have initiator domains.

achristensen07 commented on Source/WebCore/loader/ResourceLoadInfo.h:
```diff
@@ -35,12 +35,12 @@ namespace WebCore::ContentExtensions {
+static constexpr uint32_t ActionConditionMask = 0x380000;
 
 enum class ResourceType : uint32_t {
     Document = 0x0001,
```
If we are splitting this into TopDocument and ChildDocument, we should remove this and only use 1 additional bit instead of 2 additional bits.

Also, I think the addition of top-document and child-document should be split into its own PR.

xeenon replied:
Good point, then existing content blockers that still use `"document"` can parse into `TopDocument | ChildDocument`.

elijahsawyers committed:
[DNR] requestDomains/excludedRequestDomains should operate on the domain of the network request, not the domain of the loaded page
https://bugs.webkit.org/show_bug.cgi?id=289205
rdar://113878444

Reviewed by Timothy Hatcher.

This patch makes it so that requestDomains and excludedRequestDomains operate on the
domain of the network request.

To do this, we need to include the requestDomains and excludedRequestDomains in the
url-filter of the WebKit rule that we convert the dNR rule into. However, we can't
include all domains in the url-filter because the regex doesn't allow for complex
rules (such as the | operator). To get around this, we create one WebKit rule per
request domain, and we combine each request domain with the urlFilter (if there is
one). Additionally, we create one ignore-previous-rules rule for each excluded
request domain.

This patch also uses the new top-document and child-document resource types to stop
having to create multiple rules if the dNR rule doesn't specify resource types.

It's also worth noting that if a dNR rule has requestDomains and a regexFilter, we
throw away the requestDomains.

* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
(WebKit::if):
(-[_WKWebExtensionDeclarativeNetRequestRule ruleInWebKitFormat]):
(-[_WKWebExtensionDeclarativeNetRequestRule _convertRulesWithModifiedCondition:webKitActionType:chromeActionType:]):
(-[_WKWebExtensionDeclarativeNetRequestRule _webKitRuleWithWebKitActionType:chromeActionType:condition:]):
(-[_WKWebExtensionDeclarativeNetRequestRule _chromeResourceTypeToWebKitResourceType]):
(-[_WKWebExtensionDeclarativeNetRequestRule _allChromeResourceTypesForCondition:]):
(-[_WKWebExtensionDeclarativeNetRequestRule _findLongestCommonSubstringWithString:andString:]):
(-[_WKWebExtensionDeclarativeNetRequestRule _combineRequestDomain:withURLFilter:]):
(-[_WKWebExtensionDeclarativeNetRequestRule _convertedRulesForWebKitActionType:chromeActionType:]): Deleted.
(-[_WKWebExtensionDeclarativeNetRequestRule _webKitRuleWithWebKitActionType:chromeActionType:chromeResourceTypes:]): Deleted.
(-[_WKWebExtensionDeclarativeNetRequestRule _chromeResourceTypeToWebKitLoadContext]): Deleted.
(-[_WKWebExtensionDeclarativeNetRequestRule _allChromeResourceTypes]): Deleted.
Modify the dNR to WebKit rule conversion logic to work as described in the summary.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, MainFrameResourceRuleConversion)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, SubFrameResourceRuleConversion)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RepeatedMainFrameResourceRuleConversion)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, ConvertingMultipleResourceTypes)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, ConvertingXHRWebSocketAndOtherTypes)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, UpgradeSchemeForMainFrameRuleConversion)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithDomainType)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithNoSpecifiedResourceTypes)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithUnsupportedExcludedResourceTypes)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithEmptyExcludedResourceTypes)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithURLFilterAndRequestDomains)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithRequestDomains)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithExcludedRequestDomains)):
Modify the rule conversion tests to pass, and write a new test,
RuleConversionWithURLFilterAndRequestDomains, to ensure that
rules that combine request domains and URL filters converts
as expected.

Canonical link: https://commits.webkit.org/292528@main
